src/parquet/column/levels-test.cc

```diff
@@ -0,0 +1,193 @@
+    output_levels_.resize(NUM_LEVELS);
+    rle_bytes_.resize(2*NUM_LEVELS);
+    bit_packed_bytes_.resize(2*NUM_LEVELS);
+  }
+
+  void TearDown() {}
+
+ protected:
+  std::vector<int16_t> level_values_;
+  std::vector<int16_t> output_levels_;
+  LevelEncoder rle_encoder_;
+  LevelDecoder rle_decoder_;
+  LevelEncoder bit_packed_encoder_;
+  LevelDecoder bit_packed_decoder_;
+  std::vector<uint8_t> rle_bytes_;
+  std::vector<uint8_t> bit_packed_bytes_;
+  int num_rle_levels_;
+  int num_bit_packed_levels_;
+};
+//test levels with bit_width = 1
+//[0,0,0,...1,1,1,..]
+TEST_F(TestLevels, TestLevelsRle1) {
+  int max_level;
+  //generate levels
+  for (int i = 0; i < NUM_LEVELS; i++) {
+    level_values_[i] = i/(NUM_LEVELS/2);
+  }
+  max_level = 1;
+  //leave space to write the rle length value
+  rle_encoder_.Init(parquet::Encoding::RLE, max_level, NUM_LEVELS,
+       rle_bytes_.data() + sizeof(int32_t), sizeof(uint8_t)*NUM_LEVELS);
+
+  num_rle_levels_ = rle_encoder_.Encode(NUM_LEVELS, level_values_.data());
+  int32_t rle_length = rle_encoder_.getRleLength();
+  (reinterpret_cast<uint32_t*>(rle_bytes_.data()))[0] = rle_length;
+
+  ASSERT_EQ(NUM_LEVELS, num_rle_levels_);
+
+  rle_decoder_.Init(parquet::Encoding::RLE, max_level, NUM_LEVELS,
+       rle_bytes_.data());
+  int num_values = rle_decoder_.Decode(NUM_LEVELS, output_levels_.data());
+
+  ASSERT_EQ(NUM_LEVELS, num_values);
+
+  for (int i = 0; i < NUM_LEVELS; i++) {
+    ASSERT_EQ(level_values_[i], output_levels_[i]);
+  }
+
+  rle_bytes_.clear();
+  level_values_.clear();
+  output_levels_.clear();
+}
+
+//test levels with bit_width = 1
+//[0,0,0,...1,1,1,..]
+TEST_F(TestLevels, TestLevelsBitPacked1) {
+  int max_level;
+  //generate levels
+  for (int i = 0; i < NUM_LEVELS; i++) {
+    level_values_[i] = i/(NUM_LEVELS/2);
+  }
+  max_level = 1;
+  bit_packed_encoder_.Init(parquet::Encoding::BIT_PACKED, max_level, NUM_LEVELS,
+         bit_packed_bytes_.data(), sizeof(uint8_t)*NUM_LEVELS);
+  num_bit_packed_levels_ = bit_packed_encoder_.Encode(NUM_LEVELS,
+      level_values_.data());
+
+  ASSERT_EQ(NUM_LEVELS, num_bit_packed_levels_);
+
+  bit_packed_decoder_.Init(parquet::Encoding::BIT_PACKED, max_level, NUM_LEVELS,
+       bit_packed_bytes_.data());
+  int num_values = bit_packed_decoder_.Decode(NUM_LEVELS, output_levels_.data());
+
+  ASSERT_EQ(NUM_LEVELS, num_values);
+
+  for (int i = 0; i < NUM_LEVELS; i++) {
+    ASSERT_EQ(level_values_[i], output_levels_[i]);
+  }
+
+  bit_packed_bytes_.clear();
+  level_values_.clear();
+  output_levels_.clear();
+}
+
+//test levels with maximum bit-width from 2 to 8
+//[1,1,1,...3,3,3,..,7,7,7...,255,255,255]
+TEST_F(TestLevels, TestLevelsRle2to8) {
+  //for each maximum bit-width
+  for (int bit_width = 2; bit_width <= 8; bit_width++) {
+    int value = 2;
+    //generate levels uniformly upto the maximum bit-width
+    for (int i = 0; i < NUM_LEVELS; i++) {
+      level_values_[i] = ((value << (i/(NUM_LEVELS/bit_width))) - 1);
```

Comment:
Please let me know if this isn't clear. I'm looking to see data that is clearly like

`[0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, ...]`

and so forth, where the actual levels are chosen so that they fit in the desired bit width.